python/lsst/dax/data_generator/timingdict.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
+
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import time
+
+
+class TimingDict:
+    """A dictionary used to store timing information
+
+    Parameters
+    ----------
+    timing_dict : dictionary
+        A dicitionary with str keys and float values. The dictionary is
+        used internally by the class. If it is None, an empty dictionary
+        is created.
+
+    Notes
+    -----
+    While this is intended to store timing information, the only hard
+    requirement is that all the keys are strings and all the values are
+    floats.
+    The key "count" is reserved for the total count.
+    """
+
+    def __init__(self, timing_dict=None):
+        self.timing_dict = timing_dict
+        if self.timing_dict is None:
+            self.reset()
+
+    def __repr__(self):
+        return str(self.timing_dict.items())
+
+    def reset(self):
+        """Clear the dictionary
+        """
+        self.timing_dict = {}
+
+    def add(self, key, val):
+        """Add val to the corresponding key.
+
+        Parameters
+        ----------
+        key : str
+            The key name.
+        val : float
+            Value to add to the existing value. Existing value is 0 for
+            undefined keys.
+        """
+        if key in self.timing_dict:
+            self.timing_dict[key] += float(val)
+        else:
+            self.timing_dict[key] = float(val)
+
+    def start(self):
+        """Return a float start_time.
+        """
+        return time.time()
+
+    def end(self, key, start_time):
+        """Add a timing entry for 'key'
+
+        Parameters
+        ----------
+        key : str
+            Key the measurment should be added to.
+        start_time : float
+            Starting time for a measurent.
+        """
+        end_time = time.time()
+        val = end_time - start_time
+        self.add(key, val)
+
+    def increment(self):
+        """Increment "count"
+        """
+        self.add("count", 1.0)
```

Comment:
Mostly, it's easier to transmit from client to server. It may be useful to expand the concept to have different counts like "count_neigh" and have sub elements "_neigh_{whatever}" for different averages. Thoughts?

Reply:
Hmm the root of this problem really lies with doing custom serialization, because now it either pushes complexity into that serialization to handle details of the objects, or it makes the objects more complex to use so they can be simpler to serialize. Can pickling be used instead? It would make it a lot easier if it could at least handle nested dictionaries. At the very least, I'd make the serialization code responsible for separating out "count" keys from the timing keys instead of leaving them in the same dictionary. I don't really understand the relationship between what is being counted vs. the timings stored in different keys, so I can't evaluate proposals re: "count_neigh" etc.

I know this is just a tiny timing class and I'd hate for it to take a ton of effort, but I can easily imagine wanting to extend the network communication further and we'll just run into the same problems again.

Reply:
Pickling sounds like a good idea.

Reply:
Pickling has issues with data formats and it looks like it is really mean to write things to disk. count is now its own member variable.